### PR TITLE
[stdlib] Add `**` and `**=` for exponentiation

### PR DESCRIPTION
**Pull request by benrimmington**
<https://forums.swift.org/t/exponentiation-operator-and-precedence-group/44895>

Follow-up to: apple/swift#36041 and SR-14095

### CODE REVIEW

**Comment by benrimmington** [comment]
@swift-ci Please test

**Comment by benrimmington** [comment]
@swift-ci Please test Windows platform

**Comment by xwu** [comment]
Neat. I think the question (for others to decide) is whether we'd _want_ to add these shims to the standard library now (they'd have to stay forever). Personally, I think it's fine. Will have to think more about this whole precedence issue a little more on the forums.

File: test/stdlib/FloatingPoint.swift.gyb [hunk 1]
```diff
@@ -118,17 +117,17 @@ FloatingPoint.test("BinaryFloatingPoint/genericIntegerConversion") {
 FloatingPoint.test("BinaryFloatingPoint/genericFloatingPointConversion") {
   func convert<
     T: BinaryFloatingPoint, U: BinaryFloatingPoint
-  >(exactly value: T, to: U.Type) -> U { U(exactly: value) }
+  >(exactly value: T, to: U.Type) -> U? { U(exactly: value) }
```
Comment (xwu) [inline test/stdlib/FloatingPoint.swift.gyb:120]: I'm confused what happened here. Did it always not compile?

File: stdlib/public/core/FloatingPointTypes.swift.gyb [hunk 2]
```diff
@@ -742,6 +742,20 @@ extension ${Self}: BinaryFloatingPoint {
...
+% else:
+    _stdlib_pow${cFuncSuffix}(lhs, rhs)
+% end
+  }
+
+  @_alwaysEmitIntoClient
+  public static func **= (_ lhs: inout Self, _ rhs: Self) {
+    lhs = lhs ** rhs
+  }
+
```
Comment (xwu) [inline stdlib/public/core/FloatingPointTypes.swift.gyb:758]: The tricky thing about adding these for real (and @stephentyrone has said this in the forums) is that we then need to talk about whether the `(Self, Self) -> Self` overloads should vend IEEE `pow` (integral exponents get special treatment) or `powr` (derived by considering only `exp(rhs * log(lhs))`).

We would also need to consider if `(Self, Int) -> Self` overloads need to be added at the same time (vending IEEE `pown`), and if so whether they can appropriately be spelled with the same operator for two subtly different functions. The reason it'd have to be talked about now is, if we do decide to overload, then for an integer literal exponent adding the overload later would silently change the behavior of existing code.

File: test/stdlib/FloatingPoint.swift.gyb (hunk 1 above)
Comment (benrimmington) [reply test/stdlib/FloatingPoint.swift.gyb:120]: The entire test file was disabled 2 years ago (apple/swift#23423), probably due to the `isQuietNaN` test failures.

The `genericFloatingPointConversion` test was updated 5 months ago (apple/swift#33910), but it never compiled.

File: stdlib/public/core/FloatingPointTypes.swift.gyb (hunk 2 above)
Comment (benrimmington) [reply stdlib/public/core/FloatingPointTypes.swift.gyb:758]: Swift Numerics uses [`__builtin_pow{,f,l}`][shims] for all pairs of [`pow(_:_:)`][APIs] methods:

* [`Float16: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float16%2BReal.swift#L143-L158)
* [`Float32: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float%2BReal.swift#L137-L169)
* [`Float64: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Double%2BReal.swift#L150-L196)
* [`Float80: Real`](https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/Float80%2BReal.swift#L132-L144)

Clang also has:

* [`__builtin_powi{,f,l}`][powi] operations, which may be IEEE `pown`?
* `__builtin_powf16` and `__builtin_sqrtf16`, and some support for [`_Float16`][half].

[shims]: <https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/_NumericsShims/include/_NumericsShims.h#L219-L221>
[APIs]:  <https://github.com/apple/swift-numerics/blob/6583ac70c326c3ee080c1d42d9ca3361dca816cd/Sources/RealModule/ElementaryFunctions.swift#L285-L301>
[powi]:  <https://github.com/apple/llvm-project/blob/16f47cd20b1e573e25a35cce82b5d68138c0b811/clang/include/clang/Basic/Builtins.def#L165-L172>
[half]:  <https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point>

Comment (xwu) [reply stdlib/public/core/FloatingPointTypes.swift.gyb:758]: Well I think that's settled then...

Comment (benrimmington) [reply stdlib/public/core/FloatingPointTypes.swift.gyb:758]: **Correction:** Swift Numerics uses a `guard` statement, to implement `pow(x, y)` as `exp(y * log(x))`:

```swift
extension Double: Real {

  public static func pow(_ x: Double, _ y: Double) -> Double {
    guard x >= 0 else { return .nan }
    return libm_pow(x, y)
  }
}
```

So I think Clang/C99 is IEEE `pow`? Swift Numerics is IEEE `powr` and `pown`?

| Module     |                       Test | Result | IEEE   |
|:-----------|---------------------------:|-------:|:-------|
| Builtin    | `__builtin_pow(-2.0, 4.0)` | `16.0` | `pow`  |
| Darwin     |           `pow(-2.0, 4.0)` | `16.0` | `pow`  |
| RealModule |    `Double.pow(-2.0, 4.0)` | `.nan` | `powr` |
| RealModule |    `Double.pow(-2.0, 4__)` | `16.0` | `pown` |

Comment (xwu) [reply stdlib/public/core/FloatingPointTypes.swift.gyb:758]: Yeah, conceptually, that would be how they're aligned then (there are some edge conditions where the IEEE-specified result of `pow` and `powr` differ (`pow(1, .infinity)` being one example)).

**Comment by benrimmington** [comment]
I've opened apple/swift#36072 to reinstate the floating-point tests.

**Pull request closed without merging**